DevModeInitializer.getBaseDirectoryFallback() fails under Windows

### Description of the bug

Starting an application from eclipse under Windows leads to a failure in code committed with
https://github.com/vaadin/flow/pull/16266.
In our case the newly introduced code
```
            URL url = DevModeInitializer.class.getClassLoader()
                    .getResource(".");
            if (url != null && url.getProtocol().equals("file")) {
                // URI decodes the path so that e.g. " " works correctly
                String path = url.toURI().getPath();
                if (path.endsWith("/target/classes/")) {
                    return path.replaceFirst("/target/classes/$", "");
                }
            }
```
returns the following path: `/C:/projects/platform-gear-13/runtime/refimpl/frontend`
which then can not be converted in DevModeUsageStatistic:92 :
`String projectId = ProjectHelpers.generateProjectId(projectFolder);`
because it cannot be converted by
`Paths.getPath(String)`
under Windows.

The exception is
`Caused by: java.nio.file.InvalidPathException: Illegal char <:> at index 2: /C:/projects/platform-gear-13/runtime/refimpl/frontend`




### Expected behavior

It should work.

### Minimal reproducible example

Start an application in eclispe under Windows

### Versions

- Vaadin / Flow version: 23.3.11/23.37
- Java version: 11
- OS version: Windows
- Browser version (if applicable): not relevant
- Application Server (if applicable): not relevant
- IDE (if applicable): Eclipse

--- a/flow-server/src/main/java/com/vaadin/flow/server/AbstractConfiguration.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/AbstractConfiguration.java
@@ -17,13 +17,11 @@ package com.vaadin.flow.server;
 
 import java.io.File;
 import java.io.Serializable;
-import java.net.URL;
 import java.nio.file.Path;
 import java.nio.file.Paths;
 
-import org.slf4j.LoggerFactory;
-
 import com.vaadin.flow.internal.hilla.EndpointRequestUtil;
+import com.vaadin.flow.server.frontend.FileIOUtils;
 import com.vaadin.flow.server.frontend.FrontendUtils;
 
 import static com.vaadin.flow.server.InitParameters.SERVLET_PARAMETER_DISABLE_XSRF_PROTECTION;
@@ -190,48 +188,36 @@ public interface AbstractConfiguration extends Serializable {
         }
 
         String folder = getStringProperty(FrontendUtils.PROJECT_BASEDIR, null);
-        if (folder == null) {
-            /* Try determining the project folder from the classpath. */
-            try {
-                URL url = getClass().getClassLoader().getResource(".");
-                if (url != null && url.getProtocol().equals("file")) {
-                    // URI decodes the path so that e.g. " " works correctly
-                    String path = url.toURI().getPath();
-                    if (path.endsWith("/target/classes/")) {
-                        folder = path.replaceFirst("/target/classes/$", "");
-                    }
-                }
-            } catch (Exception e) {
-                LoggerFactory.getLogger(getClass()).warn(
-                        "Unable to determine project folder using classpath",
-                        e);
-            }
+        if (folder != null) {
+            return new File(folder);
         }
 
-        if (folder == null) {
-            /*
-             * Accept user.dir or cwd as a fallback only if the directory seems
-             * to be a Maven or Gradle project. Check to avoid cluttering server
-             * directories (see tickets #8249, #8403).
-             */
-            String baseDirCandidate = System.getProperty("user.dir", ".");
-            Path path = Paths.get(baseDirCandidate);
-            if (path.toFile().isDirectory()
-                    && (path.resolve("pom.xml").toFile().exists() || path
-                            .resolve("build.gradle").toFile().exists())) {
-                return path.toAbsolutePath().toFile();
-            } else {
-                throw new IllegalStateException(String.format(
-                        "Failed to determine project directory for dev mode. "
-                                + "Directory '%s' does not look like a Maven or "
-                                + "Gradle project. Ensure that you have run the "
-                                + "prepare-frontend Maven goal, which generates "
-                                + "'flow-build-info.json', prior to deploying your "
-                                + "application",
-                        path.toString()));
-            }
+        File projectFolder = FileIOUtils.getProjectFolderFromClasspath();
+        if (projectFolder != null) {
+            return projectFolder;
         }
-        return new File(folder);
+
+        /*
+         * Accept user.dir or cwd as a fallback only if the directory seems to
+         * be a Maven or Gradle project. Check to avoid cluttering server
+         * directories (see tickets #8249, #8403).
+         */
+        String baseDirCandidate = System.getProperty("user.dir", ".");
+        Path path = Paths.get(baseDirCandidate);
+        if (path.toFile().isDirectory()
+                && (path.resolve("pom.xml").toFile().exists()
+                        || path.resolve("build.gradle").toFile().exists())) {
+            return path.toAbsolutePath().toFile();
+        } else {
+            throw new IllegalStateException(String.format(
+                    "Failed to determine project directory for dev mode. "
+                            + "Directory '%s' does not look like a Maven or "
+                            + "Gradle project. Ensure that you have run the "
+                            + "prepare-frontend Maven goal, which generates "
+                            + "'flow-build-info.json', prior to deploying your "
+                            + "application",
+                    path.toString()));
+        }
     }
 
     /**

--- a/flow-server/src/main/java/com/vaadin/flow/server/frontend/FileIOUtils.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/frontend/FileIOUtils.java
@@ -17,7 +17,10 @@ package com.vaadin.flow.server.frontend;
 
 import java.io.File;
 import java.io.IOException;
+import java.net.URISyntaxException;
+import java.net.URL;
 import java.nio.charset.StandardCharsets;
+import java.nio.file.Path;
 import java.util.List;
 import java.util.stream.Collectors;
 
@@ -88,6 +91,37 @@ public class FileIOUtils {
             return null;
         }
         return FileUtils.readFileToString(file, StandardCharsets.UTF_8);
+    }
+
+    /**
+     * Try determining the project folder from the classpath.
+     *
+     * @return A file referring to the project folder or null if the folder
+     *         could not be determined
+     */
+    public static File getProjectFolderFromClasspath() {
+        try {
+            URL url = FileIOUtils.class.getClassLoader().getResource(".");
+            if (url != null && url.getProtocol().equals("file")) {
+                return getProjectFolderFromClasspath(url);
+            }
+        } catch (Exception e) {
+            log().warn("Unable to determine project folder using classpath", e);
+        }
+        return null;
+
+    }
+
+    static File getProjectFolderFromClasspath(URL rootFolder)
+            throws URISyntaxException {
+        // URI decodes the path so that e.g. " " works correctly
+        // Path.of makes windows paths work correctly
+        Path path = Path.of(rootFolder.toURI());
+        if (path.endsWith(Path.of("target", "classes"))) {
+            return path.getParent().getParent().toFile();
+        }
+
+        return null;
     }
 
 }

--- a/flow-server/src/test/java/com/vaadin/flow/server/frontend/FileIOUtilsTest.java
+++ b/flow-server/src/test/java/com/vaadin/flow/server/frontend/FileIOUtilsTest.java
@@ -1,0 +1,35 @@
+package com.vaadin.flow.server.frontend;
+
+import java.io.File;
+import java.net.URL;
+
+import org.junit.Assert;
+import org.junit.Assume;
+import org.junit.Test;
+
+import com.vaadin.open.OSUtils;
+
+public class FileIOUtilsTest {
+
+    @Test
+    public void projectFolderOnWindows() throws Exception {
+        Assume.assumeTrue(OSUtils.isWindows());
+
+        URL url = new URL(
+                "file:/C:/Users/John%20Doe/Downloads/my-app%20(21)/my-app/target/classes/");
+        Assert.assertEquals(new File(
+                "C:\\Users\\John Doe\\Downloads\\my-app (21)\\my-app\\target\\classes"),
+                FileIOUtils.getProjectFolderFromClasspath(url));
+    }
+
+    @Test
+    public void projectFolderOnMacOrLinux() throws Exception {
+        Assume.assumeFalse(OSUtils.isWindows());
+
+        URL url = new URL(
+                "file:/Users/John%20Doe/Downloads/my-app%20(21)/my-app/target/classes/");
+        Assert.assertEquals(
+                new File("/Users/John Doe/Downloads/my-app (21)/my-app"),
+                FileIOUtils.getProjectFolderFromClasspath(url));
+    }
+}

